homeassistant/components/notify/legacy.py

```diff
@@ -114,7 +115,10 @@ async def async_platform_discovered(
         """Handle for discovered platform."""
         await async_setup_platform(platform, discovery_info=info)
 
-    discovery.async_listen_platform(hass, DOMAIN, async_platform_discovered)
+    # Avoid setting up duplicate listeners after a reload
+    if not hass.data.setdefault(NOTIFY_DISCOVERY, False):
+        discovery.async_listen_platform(hass, DOMAIN, async_platform_discovered)
```

Comment:
I suggest we update the discovery helper to return the unsubscribe callback. Then we can unsubscribe when resetting the platform.

Reply:
Good idea.

Reply:
We can add a mock platform in the test.

https://github.com/home-assistant/core/blob/40d4495ed098624a1f0816357b260bd8a298d969/tests/common.py#L639

Reply:
Another issue is that the platform is setup multiple times, but we only need one dispatcher. When reloading, `async_reset_platform` is called, and the `integration` is given as a parameter. This is probably not the place to clean it up. I suppose I need to catch the signal of unloading an integration, then have a look that it is the last integration to onload, and after that call the dispatcher object to clear it.

Reply:
`async_setup` is only called once by the core. It's only called again if the the integration is reloaded.

I'm not following your point completely.

Reply:
It is about your first comment. It makes no sense to unsubscribe the callback at async_reset_platform or it will make it very complex. A reload is performed for every integration, followed by a setup.

Reply:
It seems this has been addressed with the check in `async_reset_platform`. Is there still something you want to change?

Reply:
See the last commit. Added a test as well.